Speed up compilation of common uses of std::visit()

higher-performance opened the pull request:
`std::visit` on my machine costs roughly **10 milliseconds per unique invocation** to compile, measurable as follows:

```
#include <variant>

int main(int argc, char* argv[]) {
  std::variant<char, unsigned char, int> v;
  int n = 0;
#define X(V) \
  ++n;       \
  std::visit([](int) {}, V)
#ifdef NEW_VERSION
  // clang-format off
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
  X(v); X(v); X(v); X(v); X(v); X(v); X(v); X(v);
// clang-format on
#else
  (void)v;
#endif
#undef X

  return n;
}
```

This PR hard-codes common cases to speed up compilation by roughly ~**~8x**~ (edit: 4x, see #213203 and https://github.com/llvm/llvm-project/pull/164196#issuecomment-3589875625) for them.

This approach is somewhat similar to https://reviews.llvm.org/D90168 which was reverted (https://github.com/llvm/llvm-project/issues/62648#issuecomment-1832315651), but uses `switch` statements rather than function pointer tables, for variants of up to 11 types.

higher-performance commented on libcxx/include/variant:
```diff
@@ -1578,11 +1578,42 @@ _LIBCPP_HIDE_FROM_ABI constexpr void __throw_if_valueless(_Vs&&... __vs) {
-  using __variant_detail::__visitation::__variant;
-  std::__throw_if_valueless(std::forward<_Vs>(__vs)...);
-  return __variant::__visit_value(std::forward<_Visitor>(__visitor), std::forward<_Vs>(__vs)...);
+#    define _XDispatchIndex(_I)                                                                                        \
```
Yeah, I deliberately avoided doing more work to make it easier to review and get feedback first. It adds edge cases to testing so I'm not sure how folks feel about it.

higher-performance replied:
@philnik777 is this approach something you'd be okay with?

higher-performance replied:
@frederick-vs-ja would you mind reviewing this if @philnik777 isn't reviewing?

philnik777 commented on libcxx/include/variant:
```diff
@@ -1578,11 +1582,42 @@ _LIBCPP_HIDE_FROM_ABI constexpr void __throw_if_valueless(_Vs&&... __vs) {
+                              std::integral_constant<size_t, variant_npos>,
+                              __variant_size>::value < _XDispatchCount) {
+    using __variant_detail::__access::__variant;
+    const size_t __indexes[] = {
```
```suggestion
    const size_t __indices[] = {
```

philnik777 reviewed (commented):
IIUC this basically fixes #62648. I'm not sure how to proceed here. Previously there were problems with this approach generating a lot of code, which is why it was reverted. This patch is a lot more conservative at least. I guess I'd really like if we had reflection here, since that would allow us to generate the perfect switch/case for any variant. I think I'd be fine with this as a temporary solution. I'd really like you to check compile time and code generation overhead of this though, since, as mentioned, this was a big problem previously.
Please run the variant benchmarks as well and share the results.

higher-performance committed:
Speed up compilation of std::visit() by hard-coding the most common cases

higher-performance committed:
Speed up compilation of std::visit() by hard-coding the most common cases

Somewhat similar to https://reviews.llvm.org/D90168 which was reverted (https://github.com/llvm/llvm-project/issues/62648#issuecomment-1832315651), but uses `switch` statements rather than function pointer tables, for variants of up to 11 types.

higher-performance commented:
@philnik777 bump

higher-performance commented:
@philnik777 it's been a few months. Any chance you could take a look at this?

ldionne commented on libcxx/include/variant:
```diff
@@ -1332,6 +1332,10 @@ private:
   friend struct __variant_detail::__visitation::__variant;
```
First of all, I compared with the approach that had originally been taken in https://reviews.llvm.org/D90168 (which was reverted, see https://github.com/llvm/llvm-project/issues/62648#issuecomment-1832315651 for history). Looks like this is very similar to the original patch, except it creates a switch only up to 11 variant alternatives.

Could you please update the PR description to mention https://reviews.llvm.org/D90168, mention you're using a switch statement (instead of a function ptr table), and mention the limit proposed by this patch? Just for completeness.

higher-performance replied:
Done.

ldionne commented on libcxx/include/variant:
```diff
@@ -1332,6 +1332,10 @@ private:
 };
 
+template <class... _Types>
+variant<_Types...> __upcast_to_variant(const volatile variant<_Types...>*);
```
Is this basically a way to check whether you have exactly one `_Vs...` from within `visit`? If I understand correctly and the only goal is to guard the optimization against more than one variant being visited, could we instead do this below?

```c++
using __first_variant_type = __type_pack_element<0, _Vs...>; // always at least one element in the pack
if constexpr (sizeof...(_Vs) == 1 && variant_size_v<__first_variant_type> <= _LIBCPP_VARIANT_DISPATCH_COUNT) {
  ...
}
```

higher-performance replied:
I think the upcast is important. Otherwise, you end up with `__vs.__first_variant_type::index()` being called. But in that case `__first_variant_type` is not necessarily a `std::variant`; it might be a derived class, causing the wrong `index()` method to get called. I'll add a comment to clarify.

philnik777 replied:
Does a test fail if this is removed? If not, could you add one?

higher-performance replied:
I believe one of the tests with `EvilVariantBase` failed on CI when I tried this, but it's been almost a year so I don't remember the details. I don't have a setup handy to run the tests right now unfortunately.

ldionne replied:
I checked out the code and tried a couple of ways to simplify the metaprogramming, and after 15 minutes of hitting complications, I think the approach in this patch is ~acceptable~ the best we'll get, if not super pretty. It turns out to be quite tricky to balance code duplication, correctness, forwarding and the handling of a derived classes.

ldionne commented on libcxx/include/variant:
```diff
@@ -1589,23 +1593,100 @@ _LIBCPP_HIDE_FROM_ABI constexpr void __throw_if_valueless(_Vs&&... __vs) {
+      }                                                                                                                \
+      [[__fallthrough__]]
+    switch ((..., __vs.__variant_type::index())) {
+      enum : size_t { _I0 = __LINE__ + 1 };           // New line required
```
Very clever usage of `__LINE__`. I'd add a comment like `// use __LINE__ to avoid hardcoding numbers that might become stale`

higher-performance replied:
Done.

ldionne reviewed (approved):
I have a few comments, but this basically LGTM. If you agree with my comments and are able to apply them without much difference, feel free to just land this once the CI is passing.

If you feel anything requires additional discussion, ping me on Discord to ensure this doesn't go unnoticed again.

Thanks for the patch, this actually improves the performance on AppleClang which doesn't have the Clang optimization to turn these tables into switches automatically.
